### ISSUE TITLE
Estimating unique users

### ISSUE THREAD

**Issue by hadley** [Ironholds/webtools#5]
What's state of the art on estimating number of unique users from ip addresses?

**Comment by Ironholds** [comment]
That's an interesting one. As it happens, I uh. Sort of wrote the state of the art on this for my research around session reconstruction.

(Caveat: this is with the data hitting a top-10 website, so some of the issues may not appear on lower-scale systems)

The best way of identifying uniques is to count the tuples of whatever pseudo-unique information you have; most of the time this is going to consist of IP and user agent, and occasionally accept_language, which doesn't actually add much entropy. In the case of say, the CRAN logs, we can probably safely include R version.

However, this is still not particularly great. There's still some clustering going on that screws with unique counts and session reconstruction, and the data accuracy degrades massively after around 24 hours thanks to the combination of Chrome/Firefox insisting on updating all the darn time (semantic versioning: not a thing) and dynamic IP reallocation. Again, this is mostly a problem with high-volume systems with a wide reach; if we're looking at say, traffic from universities I wouldn't necessarily expect that same degradation because they tend to use static IPs and the user doesn't have control over system updates. But you'd also get a lot more standardisation on what browser is in use, so it's a mixed bag; less accuracy, but less degradation.

Basically if you're not using a UUID cookie or a really low volume system, good luck :(.
